% IEEE Paper Template for A4 Page Size (V1)
% Sample Conference Paper using IEEE LaTeX style file for A4 pagesize.
% Copyright (C) 2006-2008 Causal Productions Pty Ltd.
% Permission is granted to distribute and revise this file provided that
% this header remains intact.
%
% REVISION HISTORY
% 20080211 changed some space characters in the title-author block
%
\documentclass[a4paper]{spie}

\usepackage[utf8]{inputenc}
\usepackage{fontenc}
\usepackage[T1]{fontenc}
\usepackage{url}
\usepackage{multirow}
\usepackage{multicol}
\usepackage{algorithm}
\usepackage{algpseudocode}
\usepackage{graphicx}
\usepackage{import}
\usepackage{xcolor}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
%\usepackage{amsthm}
\usepackage{natbib}

\newcommand{\chg}[1]{{\color{blue} #1}}
\newcommand{\todo}[1]{{\color{red} TODO: #1}}
\newcommand{\appName}{CuGene}
\newcommand{\appShortcut}{CG}

\setcitestyle{numbers,square}

%% \usepackage{lineno}
%% \pagewiselinenumbers


\title{\appName{} as a tool to view and explore genomic data}

\author{Michał Haponiuk$^{1}$, Magdalena Pawełkowicz$^{2}$, Zbigniew Przybecki$^{2}$, Robert M. Nowak$^{1}$,
  \skiplinehalf
  $^{1}$ Institute of Computer Science, Warsaw University of Technology; Nowowiejska 15/19, 00-665, Warsaw, Poland\\
  $^{2}$ Department of Plant Genetics, Breeding and Biotechnology, Warsaw University of Life Sciences; Nowoursynowska 159, 02-776, Warsaw, Poland\\
}

\authorinfo{Further author information: (Send correspondence to Robert Nowak)\\
  Robert Nowak: E-mail: r.m.nowak@elka.pw.edu.pl}

\begin{document}
\maketitle

\begin{abstract}
  Integrated \appName{} is an easy-to-use, open-source, on-line tool that can be used to browse, analyze,
  and query genomic data and annotations.
  It places annotation tracks beneath genome coordinate positions, allowing rapid visual correlation of different types of information.
  It also allows users to upload and display their own experimental results or annotation sets.
  An important functionality of the application is a possibility to find similarity between sequences by applying four different algorithms of different accuracy.
  The presented tool was tested on real genomic data and is extensively used by Polish Consortium of Cucumber Genome Sequencing.
\end{abstract}

\keywords{web application, genomic viewer, similarity finding, annotation, contig, scaffold}

\section{Introduction}

Genomic data is being uncovered in large volumes within the diverse genome sequencing projects and other new experimental technologies in molecular biology.
We now live in the post genomic era, where large data sets are produced day by day, by the sequencing technology of genomes.
Computational biology and bioinformatics provides provide computer-based methods for coping with and interpreting the genomic data.
There is a strong demand for immediate solutions in the field,
because the uncovered genomic data encode many biological insights whose deciphering can be the basis for scientific and economic success.

Genome browser is a computer program with graphical interface to browse, search, retrieve and analyze genomic sequence and annotation data.
Here we present the new genome browser named \appName{} (\appShortcut).
We attempt to give an overview of the main functions and features of this web-based tool,
covering data visualization, retrieval, analysis and customization.
To give a brief introduction to our computer program, we describe the user interface and the most important features
using the North European cucumber genome \cite{woycicki2011genome}, as an example.

\section{Software functionality}

\appName{} (\appShortcut) is a genome browser tool for visualizing and manually annotating whole genomes.
This tool provides:
\begin{itemize}
\itemsep0em
\item display of sequence and annotation tracks;
\item interactive zoom from megabase to nucleotide resolution;
\item editing of individual features, supporting manual annotation;
\item intuitive aligned sequences, elegant to view;
\item an option for being run as online client-server as well as standalone application;
\item the possibility to integrate third party applications with \appShortcut{} via HTTP API.
\end{itemize}

Though other genome browsers, listed in Tab.~\ref{tab:comparison}, with similar feature sets exist,
we believe that the presented application provides a more flexible and intuitive user interface.
This software is freely available at \url{https://github.com/mhaponiu/przegladarkaGenomow} under GNU Library or Lesser General Public License version 2.0 (LGPLv2).
The source code is open, the users with programming experience could modify and customize our application.

\appShortcut{} is platform portable, can be installed on any modern operating system (Microsoft Windows, Linux, BSD, OSx, etc.).
It was tested in Ubuntu 16.04 LTS.
No Unix or scripting knowledge is required so it is easily accessible to the average biologist with little computing experience.
The presentation layer is deployed on web browser, therefore the computer program is available for many platforms (PC, tablets, smartphones).
The graphical user interface (GUI) permits images to be drawn with minimal user input,
yet allows highly customizable figures to be generated for closer analysis or even publication.
\appShortcut{} accepts multiple sequences with annotation in standard formats (FASTA and GFF or Exel file format).
The presented computer program can handle a variety of sequences lengths, from full chromosome or scaffolds  (Fig. \ref{fig:organism})
to individual loci or genes (Fig. \ref{fig:chromosome}). Chromosome number 0 always contains faulty structures without complete information set (e.g. doesn't have start position at chromosome).
The relative orientation of each region of DNA strand (forward/reverse) can be specified.
\appShortcut{} produces an image showing genome and components such as: contigs, coding sequences (exon, introns), SNV, markers position and many others.
The user can specify an additional feature bond to specific position and/or specific region.
User-specified features connected with coloring scheme could be used to filter and to highlight interesting places or interesting regions.
Visualized genomic regions can be zoomed in and out, and scrolled left or right.
A ‘zoom’ feature enables subregions of large sequence files to be examined in more detail.

\begin{figure}[htp]
  \centering
  \includegraphics[width=0.7\textwidth]{img/chromosomy.png}
  \caption{User interface main window with cucumber chromosomes}
  \label{fig:organism}
\end{figure}


\begin{figure}[htp]
  \centering
  \includegraphics[width=0.7\textwidth]{img/skafoldy2.png}
  \caption{Detailed description of third chromosome in working window}
  \label{fig:chromosome}
\end{figure}

\section{Northern European cucumber genome}
Resarchers from Polish Consortium of Cucumber Genome Sequencing have sequenced nuclear genome of Northern European cucumber.
Consortium is the free joint-venture of scientists and businessmen from different institutions willing to cooperate in exploration of cucumber genome. The main task of the Consortium is to obtain a new knowledge from the cucumber nuclear genome sequence and implement this knowledge into amelioration of the cucumber breeding technology \cite{pawelkowicz2016next, przybecki2003isolation, przybecki2004polymorphom}. Completing the cucumber genome sequence project should be also helpful in the sequencing projects of other plant genomes.

Data about markers and contigs allowed to map sequenced genome onto chromosomes. Contigs were mapped based on Chinese markers (Ren et al. 2009) . Not every contig has been mapped successfully. Tab.\ref{tab:genome_stat} shows obtained mapping results.
General cucumber genome statistics after mapping process:\\
$\bullet$ All contigs amount: 8035 \\
$\bullet$ Mapped contigs amount: 116 \\
$\bullet$ Sum of known contig sequence lengths: 342288160 \\
$\bullet$ Sum of mapped contig sequence lengths: 196309823 \\
We have mapped around 57\% genome. Contig data includes order on chromosomes but does not have real start and stop positions. The genomic draft is a skeleton with physical and structural gaps. Contigs  are arranged in scaffolds and those using anchor markers are assigned to chromosomes. Gaps in the draft between contigs are easier to recognize thanks to experimental methods, for example, using DNA libraries. The best seems to be BAC library due it consists of large insert  between 100 – 300 kbp (Gutman et al. 2008). Sequencing of the clones from BAC library, and anchoring BES (Bac End Sequences) to the pool of contigs is basic steps in building the genome draft, because contigs could be oriented and ordered and consist the scaffold structure. Thus we can estimate the gaps across the scaffold calculating the insert length form BAC library, but this is still approximate value. More problematic are gaps between the scaffolds. In case where is no BAC library available the estmation is very difficult. Thus, to be able to visualize genome in \appName{}, we done some approximates. Between contigs there are gaps, where length of each gap depends on concrete chromosome length. In each chromosome, sum of gaps is equal to one another and sum of all mapped chromosome sequences with gaps are equal to sum of all known contigs.

\begin{table}
	\centering
	\begin{tabular}{|c||r|r|r|c|} \hline
	  \textbf{Chromosome}    & chr.\ length [Mbp] & No.\ mapped contigs & Mapped length [bp] & Mapped length [\%] \\
      \hline
      \textbf{1}             & 55                 &  22                 &  32970425          & 60\%               \\
      \textbf{2}             & 44                 &  13                 &  23992470          & 55\%               \\
      \textbf{3}             & 65                 &  14                 &  39532546          & 61\%               \\
      \textbf{4}             & 61                 &  15                 &  24781874          & 41\%               \\
      \textbf{5}             & 49                 &  22                 &  26573846          & 54\%               \\
      \textbf{6}             & 42                 &  19                 &  29507078          & 70\%               \\
      \textbf{7}             & 52                 &  11                 &  18951584          & 37\%               \\
      \hline
      $\mathbf{\Sigma}$      &368                 & 116                 & 196309823          &  57\%              \\
      \hline
	\end{tabular}
	\caption{Chromosome statistic for mapped contigs of Nothern European cucumber genome.
      The length of the chromosomes and length of the genome described in \cite{chen1998reevaluation} was used.}
	\label{tab:genome_stat}
\end{table}

\section{Architecture and implementation details}

\subsection{Architecture}

The client--server architecture was chosen to obtain high performance, the software flexibility,
its availability from a~web browser and the simplified software update.
The bioweb framework was used \cite{rn:bmri2014biosoftarch},
therefore the application is deployed on client machine (presentation layer) and on server machine (time consuming algorithms, database).
The application modules with specified interfaces are used as depicted in Fig.\ref{fig:architecture}.

The data is stored in general purpose database management system PostgreSQL.
Communication with others modules is carried out through Django Object Relational Mapping,
therefore usage of other DBMS (MySQL, SqlLite, Oracle, MsSQL etc.) is possible.
The PostgreSQL provides efficient data storage with a mechanism for the transactions.
Designed physical database model (Fig. \ref{fig:db_model}) allow us to storage different genomic structures.
Due to model is generic, user can dynamically define new types of annotations without any code interference.
It let us to define relations between annotations and create custom hierarchic tree structures. Such a solution makes application more flexible.
\appShortcut{} can storage and visualize structures without full informations about themselves (e.g. without raw sequence data).

\begin{figure}[htp]
	\centering
	\includegraphics[width=0.7\textwidth]{img/db_model.png}
	\caption{Physical database model in \appShortcut{} application.}
	\label{fig:db_model}
\end{figure}

The logic layer, responsible for data processing, involves the modules implemented in C++ and in Python.
The C++ modules include the calculation algorithms that require most computer power,
e.g. custom implementation of BLAST \cite{altschul1997gapped}, Smith-Waterman \cite{smith1981identification}, Knuth-Morris-Pratt \cite{knuth1977fast}
and Boyer-Moore \cite{boyer1977fast} methods.
The C++11 concurrency mechanism applied here speeds-up the computations in multi-core and multi-processor environment.
Communication with Django web application framework is provided by Boost.Python wrappers \cite{rn:cpp}.

The presentation layer uses a web browser (as in thin client applications).
The client partly processes information, e.g. validates user input, generates pictures to decrease the server and network load.
The client application is asynchronous -- can send and retrieve data in the background, like the Asynchronous JavaScript and XML (AJAX) techniques.
The AngularJS framework, HTML5 and the CSS Bootstrap framework are the main technologies involved in creating the client modules.

The client and server machines could be physically separated by the Internet or intranet,
because the standard HTTP server (NGINX) is used.
The messages are encapsulated of the payload in the JSON and further in HTTP format.
This will allow the data to pass through any network elements, including firewalls and address translations.

All modules are portable and could work in different operating systems and processor architectures, and only the C++ modules require the source code rebuilding.
The used libraries have non-restrictive licenses and could be used in commercial and free software.

\begin{figure}[htp]
  \centering
  \includegraphics[width=0.7\textwidth]{img/ogolny_schemat_en.png}
  \caption{Modules and its interfaces in \appShortcut{} application.
    Persistence and logic layers are deployed on server site, the presentation layer is deployed on client machine.
    All modules are portable.}
  \label{fig:architecture}
\end{figure}

\subsection{Similarity algorithms}

An important functionality of the \appShortcut{} are sequence alignment algorithms, search algorithms patterns and graphic display of alignment results.
These algorithms are particularly  applicable in matching unknown fragments of the genome sequence.
The application user has the opportunity to select one of four comparison methods: BLAST, Smith-Waterman, Knuth-Morris-Pratt and Boyer-Moore.
The example, KMP searching, is depicted in~Fig.~\ref{fig:search}.


\begin{figure}[htp]
  \centering
  \begin{tabular}{ll}
    \includegraphics[width=0.35\textwidth]{img/search.png}
    &
    \includegraphics[width=0.35\textwidth, height=4cm]{img/search_results.png}
  \end{tabular}
  \caption{The search for sequence using Knuth-Morris-Pratt algorithm (on left) and search results (on right).}
  \label{fig:search}
\end{figure}

BLAST is most widely used, heuristic algorithm, to search for similar regions in sequences.
\appShortcut{} allows customization of the minimum expect values, lengths and identities of BLAST hit to be displayed in the image.
Additionally, comparisons between two or more loci and use of previously generated tabular comparison files (converted to BLAST hit table format) is available.
The compared regions can also be rolled left, right or centered to their best BLAST hit.

In order to increase the BLAST efficiency, the Smith-Waterman algorithm is used at final step, with the highest similarity according to the BLAST.
The purpose (of this) is to examine the measures that are accurate alignment for the selected best matches.

The Knuth-Morris-Pratt is a linear time complexity algorithm to search for exact matching of strings, as well as the Boyer-Moore algorithm.
The last one is better in string benchmark, because it preprocesses the string being searched for, but not the string being searched in.
It is thus well-suited for applications in which the pattern is much shorter than the text or does persist across multiple searches.

The presented algorithms could be used for various analyzes and examinations.
In our opinion the Knuth-Morris-Pratt or/and Boyer-Moore can serve as a tool to find identical sequences
at the sequence of source or locate recurring subsequence.
The Smith-Waterman has higher time complexity, but it is accurate and includes mismatches(i.e., insertions and deletions).
The BLAST, as a representative of heuristic algorithms, is particularly useful for rapid cognition willingness highly similar sequences.

\section{Discussion}

Rapid development of sequencing technologies, increasing in effectiveness and simultaneously lowering the costs, resulted in genomes massive sequencing. The number of assembled genomes almost doubles each year.are growing in avalanche.
Beside information about the order of nucleotides,
DNA sequence becomes a source of data for further steps like structural and functional annotation, polymorphisms discovery, and markers detection, and very broad study of genome evolution. This involves multiple fields such as structural analysis of the genome, the study of genomic features, gene and ancient genome duplications, polyploidy, and comparative genomics. Genome evolution is a constantly changing and evolving field due to the steadily growing number of sequenced genomes and transcriptomes, both prokaryotic and eukaryotic, available to the scientific community and the public at large.
Moreover due to increase  number of sequenced data  (genomes and transcriptomes) there is a huge need to upswing their features in a simple way with maintaining more information for example as a graphical output in the browser. A complete genome sequence of an organism can be considered to be the ultimate genetic map, in the sense that the heritable characteristics are encoded within the DNA and that the order of all the nucleotides along each chromosome is known. Finding all the functional parts of genome sequences and using this information to improve our knowledge of the  organism actions is very important. The genes are the fundamental units of inheritance and the ultimate determinant of all phenotypes. According to the “central dogma” of molecular biology, the process of transcribing DNA into RNA, or translating RNA into protein are very interesting for  researchers. The process of transcription involves creating a perfect RNA copy of the gene using the DNA of the gene as a template. Every gene consists of several functional components, each involved in a different facet of the process of gene expression. In eukaryotes, the coding regions of most genes are not continuous. They consist of areas that are transcribed into mRNA, the exons, which are interrupted by stretches of DNA that do not appear in mature mRNA, the introns. Transcription is the first step of gene regulation in which particular segment of DNA is copied into pre-mRNA, then pre-mRNA is spliced \cite{chen2009mechanisms}. This is a process in which introns are removed from a mRNA precursor. Also important feature is a promoter. This is a region of DNA that initiates transcription of a particular gene. Promoters are located near the transcription start sites of genes, on the same strand and upstream on the DNA (towards the 5' region of the sense strand). The coding region and promotor region arevery important are two main functional units \cite{pawelkowicz2016next}.
Such readily accessible data have encouraged the proliferation of adaptive arguments for the evolution of gene and genomic features, often with little or no attention being given to simpler and more powerful alternative explanations. Comparative genomic and transcriptomics develops intensely mainly on bioinformatics fields so possessing browsers for visualization of compared data sets is indispensable.
In this vast accumulation of information,
a simple and clear way of visualization, access and modification of data is necessary.
Since early 2000s the web based browsers for genomic purposes have been developed.
Simultaneously, several large projects for genomic data integration and browsers implementations were executed
resulting in Ensembl \cite{hubbard2002ensembl}, UCSC \cite{karolchik2003ucsc} and Gremene \cite{ware2002gramene} web platforms.
They were started as free online services with implemented genomic and transcriptomic databases from one or many species,
enabling simplified data mining features.
Subsequently, some of them provided browsers framework.
At the same time, a framework for local usage by scientists was constructed by Generic Model Organism Database (GMOD) project \cite{stein2002generic}.
It enabled every research group to convert FASTA files and tabular data into easily accessible, interactive web page with embedded bioinformatics tools.

\appName{} was developed due to two main factors: rapidly expanding genomic database of Northern European cucumber, containing massive amounts of data,
separated in several not integrated sections impeded information access and transparency.
Secondly, most popular framework for local installation and creation genomic browsers GBrowse is based on static web pages and
does not provide smooth interactive browsing like frameworks based on modern technologies.

Our software is a web based framework for storage and management of massive genomic, transcriptomic and proteomic data.
Configuration and format of introduced information has been simplified versus most programs allowing for smoother and swifter start up for new species or varieties.
Currently, one species application is functional with a goal for multispecies genome comparisons like  Ensembl and UCSC as depicted in Tab.~\ref{tab:comparison},
however it is possible to input several genomes data for independent browsing  but without interaction mode.
The most important feature provided by all available browsers and \appShortcut{} is data input and comparison executed by end users through the user interface.
This function is essential for genomic and transcriptomic information sharing with others research teams.
Contrary to Alamut \url{http://www.interactive-biosoftware.com/}, Celera Genome Browser \url{http://sourceforge.net/projects/celeragb/}
and Vista \cite{visel2007vista}, \appShortcut{} implemented as a web browser becomes a universal tool accessible for different devices ranging from personal computers,
tablets to smartphones and compatible with both windows and UNIX based systems.
Similarly to the most developed AJAX-based JBrowse \cite{skinner2009jbrowse}, ABrowse \cite{kong2012abrowse} and Anno-J \cite{lister2008highly}
\appShortcut{} provides full interactive access to genomic data with smooth scanning throughout the whole genome,
exhibiting nucleotide sequence, structural annotation and functions of annotated genes.
Being also stand-alone framework \appShortcut{} can be used as a data sharing tool for sensitive information
among researchers from one department or institution as opposed to genome browsers implemented only for online operation.
\appShortcut{} like most browser supports search function; however, this browser exploits three different algorithms including BLAST,
which is not implemented in several modern browsers: Synbrowse \cite{pan2005synbrowse}, GeneWall \url{http://www.wobblebase.com/GeneWall},
Celera Genome Browser.
This results in more accurate and precise outcome of new genes, regulatory elements and markers localization in genome and identifying homologous
from different varieties and species.
Main processes are carried out on  high power server with a broadband internet connection,
where all databases are present and every inquiry from client is analyzed and elaborated providing fast and high quality service
even for end users with low computational power.

\begin{table}

  \centering

  \begin{tabular}{|l|c|c|c|c|c|c|c|c|c|p{1cm}|p{1cm}|} \hline
    ~ & Embedded & \multicolumn{2}{|c|}{Custom species}
    & \multicolumn{3}{|c|}{Searching} & Frame- & Stand & Mobile \\
    ~ & data & one & many
    & BLAST & SW & KMP & work & -alone & devices \\ \hline
    Ensembl               & + & + & + &  + & - & - &  + &+ & +  \\ \hline
    UCSC                  & + & + & + &  + & - & - &  + &+ & +  \\ \hline
    Gramene               & + & + & + &  + & - & - &  - &- & +  \\ \hline
    Gbrowse               & - & + & + &  + & - & - &  + &+ & +  \\ \hline
    Jbrowse               & - & + & + &  + & - & - &  + &+ & +  \\ \hline
    Abrowse               & - & + & + &  + & - & - &  + &+ & +  \\ \hline
    Anno-J                & - & + & + &  + & - & - &  + &+ & +  \\ \hline
    SynBrowse             & - & + & + &  - & - & - &  + &+ & -  \\ \hline
    SynView               & - & + & + &  - & - & - &  + &+ & +  \\ \hline
    LookSeq               & - & + & + &  + & - & - &  + &+ & +  \\ \hline
    Vista                 & + & - & - &  + & - & - &  - &- & -  \\ \hline
    Celera Genome Browser & - & + & - &  - & ND& ND&  + &+ & -  \\ \hline
    Alamut                & +*& - & - &  + & ND& ND&  - &- & -  \\ \hline
    GeneWall              & +*& - & - &  - & - & - &  - &+ & +  \\ \hline
    GenomeView            & - & + & + &  + & - & - &  + &+ & -  \\ \hline
    \textbf{\appName{}}   & - & + & + &  + & + & + &  + &+ & +  \\ \hline
  \end{tabular}

  \caption{Genome browsers functionality comparison; Embedded data -- online service with embedded data,
    Custom Species -- ability to add user data on one or many species,
    SW -- Smith-Waterman algorithm, KMP -- Knuth-Morris-Pratt algorithm,
    Framework -- ability to install software on user server, Stand-alone -- ability to install software on PC without Internet access,
    ND -- no data, * -- only human database}
  \label{tab:comparison}

\end{table}

\section{Conclusion}

The \appName{} implements most modern solutions for managing genomic resources obtained through new generation sequencing,
structural and functional annotation and more.
User friendly interface enables fast and effective data search.
Genomic and annotation data exhibit in \appShortcut{} provides a different view on it contrary to tabular management, especially with North European Cucumber.
Embedded tools extended by internet access even in stand-alone distribution simplify the work of a researcher,
who can focus on results rather then on analysis processes.
This instrument allows for swift information exchange with other research institution with a possibility of restricting sensitive date to your department.
Further development of \appName{} will be implemented to keep up with genomic and transcriptomic massive data analysis.

\acknowledgments

This work was supported by grants from National Science Center 2013/11/B/NZ/00814 and the statutory research of Institute of Computer Science of Warsaw University of Technology.

\bibliographystyle{unsrt}
\bibliography{webomics.bib}

\end{document}
